Basic handling of function pointer types in the IDE

333fred opened the pull request:
* Offer `delegate` keyword when function pointer types are permissible.
* Completion inside function pointer type argument lists (including valid ref modifiers)
* SymbolKey support.
* Basic formatting support.
* Support in add/remove parameter.

333fred commented:
@CyrusNajmabadi @jasonmalinowski I'm not publishing this yet because the previous PR isn't merged, but the relevant changes are in the last commit and are basically ready for review (sans tests). I'm going to get started on tests locally, but I'd appreciate an initial pass of the implementation.

333fred commented on src/Features/Core/Portable/Shared/Extensions/ISymbolExtensions_2.cs:
```diff
@@ -129,6 +129,9 @@ public static Glyph GetGlyph(this ISymbol symbol)
                     return ((IPointerTypeSymbol)symbol).PointedAtType.GetGlyph();
 
+                case SymbolKind.FunctionPointer:
+                    return Glyph.Intrinsic;
```
I chose a glyph that seemed appropriate here. If there's a better suggestion, please let me know :)

CyrusNajmabadi replied:
wfm

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:
```diff
@@ -1693,7 +1693,9 @@ public static NameSyntax ParseName(string text, int offset = 0, bool consumeFull
-        public static TypeSyntax ParseTypeName(string text, int offset = 0, bool consumeFullText = true)
+        // Backcompat overload, do not remove
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public static TypeSyntax ParseTypeName(string text, int offset, bool consumeFullText)
```
why is this not just deferring to the other overload?

333fred replied:
I wanted to leave existing code exactly as is. It probably _could_ do that though.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/ChangeSignature/CSharpChangeSignatureViewModelFactoryService.cs:
```diff
@@ -44,7 +44,9 @@ public override SymbolDisplayPart[] GeneratePreviewDisplayParts(AddedParameterVi
-        public override bool IsTypeNameValid(string typeName) => !SyntaxFactory.ParseTypeName(typeName).ContainsDiagnostics;
+#nullable enable
+        public override bool IsTypeNameValid(string typeName, ParseOptions? parseOptions) => !SyntaxFactory.ParseTypeName(typeName, options: parseOptions).ContainsDiagnostics;
+#nullable restore
```
file is tiny.  i would prefer to nrt all of it or none of it.

333fred replied:
I'm going to do none. It implements an interface and I don't want to go annotating multiple levels here.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/ChangeSignature/AddParameterDialogViewModel.cs:
```diff
@@ -226,7 +226,8 @@ private void SetCurrentTypeTextAndUpdateBindingStatus(string typeName)
             var languageService = Document.GetRequiredLanguageService<IChangeSignatureViewModelFactoryService>();
-            return languageService.IsTypeNameValid(typeName);
+            _ = Document.TryGetSyntaxTree(out var syntaxTree);
+            return languageService.IsTypeNameValid(typeName, syntaxTree?.Options);
```
yeah, this is really hacky.  why do we need options to know if a type name is valid?

333fred replied:
Because if the langversion isn't set appropriately, the parser will error because the feature isn't available.

CyrusNajmabadi replied:
comment left above on how i think we should do this isntead.  would def want thsi line undone.

333fred replied:
Now passing latest.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolKey/SymbolKey.FunctionPointerTypeSymbolKey.cs:
```diff
@@ -0,0 +1,43 @@
+                    return default;
+                }
+
+                using var _ = paramTypes;
```
should this not be above the .IsDefault check?

333fred replied:
It's explicitly below, because if `paramTypes` is default there's nothing to dipose.

CyrusNajmabadi replied:
docs say:

```
            /// Callers should <see cref="IDisposable.Dispose"/> the instance returned.  No check is
            /// necessary if <c>default</c> was returned before calling <see cref="IDisposable.Dispose"/>
            /// </summary>
```

CyrusNajmabadi replied:
and all existing callers put this in a `using`.  so i'd far prefer you keep your code using that pattern and keeping things simple that way.  thanks!

333fred replied:
Done.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolKey/SymbolKey.SymbolKeyWriter.cs:
```diff
@@ -298,6 +299,8 @@ internal void WriteRefKindArray(ImmutableArray<IParameterSymbol> values)
                 EndKey();
             }
 
+            internal void WriteRefKind(RefKind refKind) => WriteInteger((int)refKind);
```
why internal?

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Extensions/ISymbolExtensions_Accessibility.cs:
```diff
@@ -103,6 +103,23 @@ internal static partial class ISymbolExtensions
                     return IsSymbolAccessibleCore(((IPointerTypeSymbol)symbol).PointedAtType, within, null, out failedThroughTypeCheck);
 
+                case SymbolKind.FunctionPointer:
+                    var funcPtrSign = ((IFunctionPointerTypeSymbol)symbol).Signature;
```
```suggestion
                    var funcPtrSignature = ((IFunctionPointerTypeSymbol)symbol).Signature;
```

333fred replied:
Renamed.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxNodeExtensions.cs:
```diff
@@ -30,5 +30,7 @@ public static bool IsDelegateOrConstructorOrLocalFunctionOrMethodOrOperatorParam
             return false;
         }
+
+        public static bool IsInFunctionPointer(this SyntaxToken node) => node.Parent.IsKind(SyntaxKind.FunctionPointerType);
```
this isn't named well (And i don't think we need this extension at all).  "is in" implies to me it's in the subtree of that function-pointer, not that its' directly parented by the function pointer.  rename, or just inline this at callers.

333fred replied:
Removed.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -902,6 +903,30 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
+            {
+                case SyntaxKind.LessThanToken:
+                case SyntaxKind.CommaToken:
+                    return token.Parent is FunctionPointerTypeSyntax;
```
the comma is parented direction by the function pointer type?  i.e. we don't have FunctionPointerTypeList?  seems strangely inconsistent with the rest of hte syntactic model.

333fred replied:
It probably should be a separate parent, I'm planning on changing that as part of the rework.

CyrusNajmabadi replied:
if you have an issue tracking the "missing node" issue, can you potentially add to that?  would prefer tehse not get lost, esp. as we get into crunch time.

333fred replied:
Added a link.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -902,6 +903,30 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
+                return false;
+            }
+
+            return syntaxTree.IsInPartiallyWrittenFunctionPointer(position, cancellationToken);
```
the partially-written stuff is necessary for normal generics, which are ambiguous with expressions.  i.e. `List<a` could be an expr or a generic type.  for func-ptrs, we never have that ambiguity.  You have `delegate*` so we always know its' a funcptr.

333fred replied:
Simplified.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SyntaxTreeExtensions.cs:
```diff
@@ -92,6 +92,11 @@ public static bool IsInNonUserCode(this SyntaxTree syntaxTree, int position, Can
 
+        public static bool IsInPartiallyWrittenFunctionPointer(this SyntaxTree syntaxTree, int position, CancellationToken cancellationToken)
+        {
+            return syntaxTree.IsInPartiallyWrittenGenericOrFunctionPointer(position, lookForFunctionPointer: true, cancellationToken, out _, out _);
```
i don't think you need this.

333fred replied:
Removed.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/SyntaxTreeExtensions.cs:
```diff
@@ -164,6 +181,15 @@ public static bool IsInNonUserCode(this SyntaxTree syntaxTree, int position, Can
+                                    genericIdentifier = token;
+                                    return true;
+                                }
+                            }
```
i would undo this.

333fred replied:
Done.

CyrusNajmabadi commented on src/EditorFeatures/CSharp/AutomaticCompletion/Sessions/LessAndGreaterThanCompletionSession.cs:
```diff
@@ -35,6 +35,7 @@ public override bool CheckOpeningPoint(IBraceCompletionSession session, Cancella
             // type argument or parameter list
             if (!token.CheckParent<TypeParameterListSyntax>(n => n.LessThanToken == token) &&
                 !token.CheckParent<TypeArgumentListSyntax>(n => n.LessThanToken == token) &&
+                !token.CheckParent<FunctionPointerTypeSyntax>(n => n.LessThanToken == token) &&
```
is there test for this?

333fred replied:
Added `FunctionPointerStartSession` in `AutomaticLessAndGreaterThanCompletionTests`

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest/ExtractMethod/ExtractMethodTests.cs:
```diff
@@ -11185,5 +11185,36 @@ private static bool NewMethod()
+        _ = NewMethod(ptr1)(ref s);
+    }
+
+    private static delegate*<ref string, ref readonly int> NewMethod(delegate*<delegate*<ref string, ref readonly int>> ptr1)
```
lordy those look awful :D

333fred replied:
Honestly, I don't think it's that bad. It's very similar to `Func`. And certainly a heck of a lot better than trying to put actual method syntax in there (which doesn't feel very C# to me at all).

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest2/Recommendations/DecimalKeywordRecommenderTests.cs:
```diff
@@ -732,5 +732,14 @@ void Method()
+            await VerifyKeywordAsync(@"
+class C
+{
+    delegate*<$$");
```
would like at least one test of:
1. not getting these after `delegate*`
2. getting thse after `delegate*<int, `
3. getting these after `delegate*<ref ` (and after comma as well).

this doesn't have tobe duplciated into each file. but we should have it in at least one of them.

CyrusNajmabadi replied:
i honestly would prefer not to. i don't get 'test fatigue' personally.  the test sits there enert, doing it's duty in a trivial to understand fashion.

more complex test harnesses annoy the heck out of me as, so often,  cannot actually tell what they're doing and if they're doing it properly.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:
```diff
@@ -43,16 +44,16 @@ public async Task TestAfterGlobalVariableDeclaration_Interactive()
         {
-            await VerifyAbsenceAsync(
+            await VerifyKeywordAsync(
 @"using Goo = $$");
```
wait why?  i would still expect these to *not* be offered here.

333fred replied:
Why not? You can alias them.

CyrusNajmabadi replied:
Really?  We don't let you alias tuples, or pointers or arrays...  But we let you alias a funcptr??

333fred replied:
Hmm, I thought we did, but apparently we do not.

333fred replied:
Adjusted.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:
```diff
@@ -235,16 +236,22 @@ public async Task TestNotAfterSealed()
+        public async Task TestAfterStatic()
+        {
+            await VerifyAbsenceAsync(SourceCodeKind.Regular, @"static $$");
+            await VerifyKeywordAsync(SourceCodeKind.Script, @"static $$");
```
why absent in srtipt?

333fred replied:
They're allowed in script, not in regular. On the top level, `static` can't be a type.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:
```diff
@@ -297,25 +304,25 @@ public async Task TestInMemberInitializer1()
         {
-            await VerifyAbsenceAsync(AddInsideMethod(
+            await VerifyKeywordAsync(AddInsideMethod(
 @"typeof($$"));
```
interesting.  does that work at runtime?

333fred replied:
Gives you back `IntPtr`

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:
```diff
@@ -297,25 +304,25 @@ public async Task TestInMemberInitializer1()
         {
-            await VerifyAbsenceAsync(AddInsideMethod(
+            await VerifyKeywordAsync(AddInsideMethod(
 @"sizeof($$"));
```
does this work?

333fred replied:
Yep.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest2/Recommendations/DynamicKeywordRecommenderTests.cs:
```diff
@@ -633,5 +633,14 @@ public async Task TestAfterAsync()
+            await VerifyKeywordAsync(@"
+class C
+{
+    delegate*<$$");
```
presumably this isn't actually allowed?  or is it?  wonky!

333fred replied:
Of course it's allowed. Would kinda suck if you couldn't nest function pointers.

CyrusNajmabadi replied:
This is 'dynamic'. I'm surprised you can have a funcptr that takes a dynamic arg.

333fred replied:
You can't convert a function pointer to `dynamic`. But a function pointer is just a method signature. You can put pretty much anything you can put in a normal method signature in a function pointer.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest2/Recommendations/ReadOnlyKeywordRecommenderTests.cs:
```diff
@@ -588,5 +588,27 @@ void M()
+            await VerifyAbsenceAsync($@"
+class C
+{{
+    delegate*<{modifier} $$");
```
oh, you can't say `delegate*<readonly int x>`?  you have to say `ref readonly` only?

CyrusNajmabadi replied:
i'd honestly prefer a dedicated test for taht.

333fred replied:
Not sure I understand?

333fred replied:
Added test from offline conversation.

CyrusNajmabadi commented on src/EditorFeatures/CSharpTest2/Recommendations/VoidKeywordRecommenderTests.cs:
```diff
@@ -727,5 +727,60 @@ void M(Action<int> a)
+            await VerifyAbsenceAsync($@"
+class C
+{{
+    delegate*<int, {refKind} $$");
```
can't have `ref void**`?

333fred replied:
Thats... a good point. Will certainly make the code simpler.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/KeywordRecommenders/DelegateKeywordRecommender.cs:
```diff
@@ -30,7 +30,7 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
             return
                 context.IsGlobalStatementContext ||
-                (context.IsNonAttributeExpressionContext && !context.IsConstantExpressionContext) ||
+                ((context.IsNonAttributeExpressionContext || context.IsTypeContext) && !context.IsConstantExpressionContext) ||
```
nope.  hate it.  can't read it.  extract out into other statement with a comment.

CyrusNajmabadi commented on src/Features/Core/Portable/ExtractMethod/MethodExtractor.TypeParameterCollector.cs:
```diff
@@ -29,6 +29,15 @@ public override void VisitDynamicType(IDynamicTypeSymbol dynamicTypeSymbol)
+                foreach (var param in symbol.Signature.Parameters)
+                {
+                    param.Type.Accept(this);
+                }
```
do we have a test of a feature that goes through thisx codepath?

333fred replied:
Added a method extraction test that involves a type parameter in a function pointer.

CyrusNajmabadi commented on src/VisualStudio/CSharp/Impl/ChangeSignature/CSharpChangeSignatureViewModelFactoryService.cs:
```diff
@@ -44,7 +44,7 @@ public override SymbolDisplayPart[] GeneratePreviewDisplayParts(AddedParameterVi
         }
 
-        public override bool IsTypeNameValid(string typeName) => !SyntaxFactory.ParseTypeName(typeName).ContainsDiagnostics;
+        public override bool IsTypeNameValid(string typeName, ParseOptions parseOptions) => !SyntaxFactory.ParseTypeName(typeName, options: parseOptions).ContainsDiagnostics;
```
still confsued about this.  not sure why parseoptions are necessary to ask if a tyep name is valid.

CyrusNajmabadi replied:
note: imo, what we shoudl be doing here is removing this options, and just passing in Latest to ParseTypeName.  that way we're effectively saying "are there any diagnostics *unrelated* to "you're using teh wrong langversion".

333fred replied:
`delegate*<void>` is _syntactically_ invalid if the language version is not `Preview`. We don't have this issue with any existing types today (even `nint`) because they can't be determined to be invalid from syntax alone. I'm certainly fine with passing `Preview` as the default langversion however, as we'll appropriately error about langversion when it's actually added.

jasonmalinowski replied:
Agreed, seems best to just pass in latest. What's funky about trying to do a check here is it's still not necessarily valid everywhere _else_ the actual fix might apply to, so it's not really protecting the user from much.

333fred replied:
Now passing latest.

CyrusNajmabadi commented on src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:
```diff
@@ -200,13 +200,14 @@ public override SyntaxNode ParameterDeclaration(string name, SyntaxNode type, Sy
         }
 
-        internal static SyntaxTokenList GetParameterModifiers(RefKind refKind)
+        internal static SyntaxTokenList GetParameterModifiers(RefKind refKind, bool forFunctionPointerReturnParameter = false)
```
i don't think this new parameter has value.  ca we just remove and respect the refkind passed in.  i.e. if somone asks for `.Out` they get `out`.  c'est la vie, that's what they asked for.

333fred replied:
No. When we generate the modifiers for the return type of a function pointer, `In` means `RefReadonly`. These values are the same.

333fred replied:
I'll add a comment that `In` == `RefReadOnly`

CyrusNajmabadi replied:
I think I get it. I think I want this parameter to be more like 'bool isParameter'. In parameters we allow 'in' but not in places like a property return.

333fred replied:
I considered making it `isParameter`, but given that the function itself is used for generating parameter modifiers I felt it better to be specific about the scenario here. Regular return type modifiers are parsed and handled in a different way that parameter modifiers so I don't want to make the function name less specific.

333fred replied:
Added more to the comment. I'm keeping the parameter anme.

CyrusNajmabadi commented on src/Workspaces/CSharpTest/Formatting/FormattingTests_FunctionPointers.cs:
```diff
@@ -29,5 +29,24 @@ unsafe class C
+            var content = @"
+unsafe class C
+{
+    delegate *cdecl < int,  int> functionPointer;
```
hurgh.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/SymbolKey/SymbolKey.FunctionPointerTypeSymbolKey.cs:
```diff
@@ -0,0 +1,36 @@
+                var returnRefKind = reader.ReadRefKind();
+                var returnType = reader.ReadSymbolKey();
+                using var paramRefKinds = reader.ReadRefKindArray();
+                using var paramTypes = reader.ReadSymbolKeyArray<ITypeSymbol>();
```
i can't remember the semantics of ReadSymbolKeyArray, does it fail if any of them fail?  or does it just toss that value.  if the latter, you hve to check that the lengths of these two arrays are the same.  (which is pretty sane to do anyways).

333fred replied:
It fails if any of them fail.

CyrusNajmabadi commented on src/Workspaces/CoreTest/SymbolKeyTests.cs:
```diff
@@ -783,6 +784,21 @@ void Method((C a, int b) t)
+
+            var comp = GetCompilation(source, LanguageNames.CSharp);
+            var fields = GetDeclaredSymbols(comp).OfType<IFieldSymbol>().Select(f => f.Type);
+            TestRoundTrip(fields, comp);
```
thanks.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:
```diff
@@ -115,9 +116,11 @@ public override TypeSyntax VisitDynamicType(IDynamicTypeSymbol symbol)
+                    .SelectAsArray(t => SyntaxFactory.Parameter(SyntaxFactory.MissingToken(SyntaxKind.IdentifierToken)).WithModifiers(t.RefKindModifiers).WithType(t.Type.GenerateTypeSyntax()));
 
                 return AddInformationTo(
                     SyntaxFactory.FunctionPointerType(SyntaxFactory.SeparatedList(parameters)), symbol);
```
how do you specify the `cdecl` bit here?

333fred replied:
You can't today. That's why there's a comment above about it.

333fred commented on src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:
```diff
@@ -235,16 +236,22 @@ public async Task TestNotAfterSealed()
-        public async Task TestNotAfterDelegate()
-            => await VerifyAbsenceAsync(@"delegate $$");
+        public async Task TestAfterDelegate()
+            => await VerifyKeywordAsync(@"delegate $$");
```
Yep, you can have `delegate delegate*<void> M()`.

333fred commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -1010,6 +1042,17 @@ public static bool IsDefaultExpressionContext(this SyntaxTree syntaxTree, int po
 
+#if !CODE_STYLE
+            if (token.IsKind(SyntaxKind.CommaToken) &&
+                token.Parent.IsKind(SyntaxKindEx.FunctionPointerType, out FunctionPointerTypeSyntax funcPtrType))
```
No, that's  where we started. That needs to do a bit more work than we need to do here, as this can just be a simple parent check.

333fred commented on src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:
```diff
@@ -200,13 +200,14 @@ public override SyntaxNode ParameterDeclaration(string name, SyntaxNode type, Sy
             {
                 RefKind.None => new SyntaxTokenList(),
-                RefKind.Out => SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.OutKeyword)),
+                RefKind.Out when !forFunctionPointerReturnParameter => SyntaxFactory.TokenList(SyntaxFactory.Token(SyntaxKind.OutKeyword)),
```
Sure, I can remove it.

CyrusNajmabadi commented on src/Features/CSharp/Portable/Completion/KeywordRecommenders/DelegateKeywordRecommender.cs:
```diff
@@ -30,13 +30,15 @@ protected override bool IsValidContext(int position, CSharpSyntaxContext context
             return
                 context.IsGlobalStatementContext ||
-                ((context.IsNonAttributeExpressionContext || context.IsTypeContext) && !context.IsConstantExpressionContext) ||
+                (ValidTypeContext(context) && !context.IsConstantExpressionContext) ||
```
can you just pull in the IsConstantExpressionContext into that?  it's only checked there and for that purpose.  then this expression is a simple sequence of 'ors'.  not `this or this or or this and not this or this or this`.

333fred replied:
I actually felt this was more readable.

333fred replied:
Pulled it out and added the using directive change.

CyrusNajmabadi commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SymbolEquivalenceComparer.cs:
```diff
@@ -184,6 +184,10 @@ private static bool CheckContainingType(IMethodSymbol x)
+            else if (x.MethodKind == MethodKindEx.FunctionPointerSignature)
+            {
+                return false;
+            }
```
1. doc
2. when do you hit this?

333fred replied:
Added more docs.

333fred commented on src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:
```diff
@@ -2,6 +2,7 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Diagnostics.Tracing;
```
Removed

333fred commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/SpacingFormattingRule.cs:
```diff
@@ -320,6 +320,29 @@ public override AbstractFormattingRule WithOptions(AnalyzerConfigOptions options
+                        // No spacing between the calling convention and opening angle bracket of function pointer types:
+                        // delegate* cdecl<
+                        case SyntaxKind.IdentifierToken:
+                            return CreateAdjustSpacesOperation(0, AdjustSpacesOption.ForceSpaces);
```
Removed.

333fred commented on src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Formatting/Rules/SpacingFormattingRule.cs:
```diff
@@ -320,6 +320,29 @@ public override AbstractFormattingRule WithOptions(AnalyzerConfigOptions options
+                        // No spacing between the calling convention and opening angle bracket of function pointer types:
+                        // delegate* cdecl<
+                        case SyntaxKind.IdentifierToken:
+                            return CreateAdjustSpacesOperation(0, AdjustSpacesOption.ForceSpaces);
```
Done.

333fred commented:
@CyrusNajmabadi @jasonmalinowski ready for another look.

333fred commented:
@dotnet/roslyn-compiler @AlekseyTs for the small compiler api change change in this PR.

cston commented on src/Compilers/VisualBasic/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -1 +1,3 @@
-
+Shared Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ParseTypeName(text As String, offset As Integer = 0, options As Microsoft.CodeAnalysis.ParseOptions = Nothing, consumeFullText As Boolean = True) -> Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax
+Shared Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ParseTypeName(text As String, offset As Integer, consumeFullText As Boolean) -> Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax
+*REMOVED*Shared Microsoft.CodeAnalysis.VisualBasic.SyntaxFactory.ParseTypeName(text As String, offset As Integer = 0, consumeFullText As Boolean = True) -> Microsoft.CodeAnalysis.VisualBasic.Syntax.TypeSyntax
```
>*REMOVED* [](start = 0, length = 9)

Why is this overload marked as removed?

jasonmalinowski commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -3722,10 +3722,10 @@ unsafe void M()
                 //         Action ptr1 = (Action)&M;
                 Diagnostic(ErrorCode.ERR_BadSKunknown, "Action").WithArguments("System.Action", "type").WithLocation(8, 24),
-                // (9,23): error CS8800: Cannot convert a & method group 'M' to delegate type 'M'.
+                // (9,23): error CS8800: Cannot convert & method group 'M' to delegate type 'M'.
```
Not to bikeshed on the message too much but I admit seeing a & with spaces around it my brain just parsed as a shortcut for "and" and this reads as "cannot convert and method group" and it was pretty confusing. "Cannot convert the address of a method group 'M'" or consider removing the space after & to make it clear it's the operator?

333fred replied:
Done

jasonmalinowski commented on src/EditorFeatures/CSharpTest2/Recommendations/DelegateKeywordRecommenderTests.cs:
```diff
@@ -43,16 +43,25 @@ public async Task TestAfterGlobalVariableDeclaration_Interactive()
 
         [Fact, Trait(Traits.Feature, Traits.Features.KeywordRecommending)]
-        public async Task TestNotInUsingAlias()
+        public async Task TestInUsingAlias()
```
Test name still needs to be updated back to the original name?

333fred replied:
Fixed.

jasonmalinowski commented on src/VisualStudio/VisualBasic/Impl/ChangeSignature/VisualBasicChangeSignatureViewModelFactoryService.vb:
```diff
@@ -41,7 +41,8 @@ Namespace Microsoft.VisualStudio.LanguageServices.VisualBasic.ChangeSignature
 
         Public Overrides Function IsTypeNameValid(typeName As String) As Boolean
-            Return Not SyntaxFactory.ParseTypeName(typeName).ContainsDiagnostics
+            Static visualBasicParseOptions As New VisualBasicParseOptions(LanguageVersion.Latest)
```
Probably should have a similar comment to the C# one.

333fred replied:
I don't think it really matters as much for the VB version. We're not adding new syntax, and VB doesn't have a concept of `Preview`.

jasonmalinowski commented on src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/Extensions/ITypeSymbolExtensions.TypeSyntaxGeneratorVisitor.cs:
```diff
@@ -115,9 +118,11 @@ public override TypeSyntax VisitDynamicType(IDynamicTypeSymbol symbol)
-                var parameters = symbol.Signature.Parameters.Select(p => p.Type)
-                    .Concat(SpecializedCollections.SingletonEnumerable(symbol.Signature.ReturnType))
-                    .SelectAsArray(t => SyntaxFactory.Parameter(SyntaxFactory.MissingToken(SyntaxKind.IdentifierToken)).WithType(t.GenerateTypeSyntax()));
+                var parameters = symbol.Signature.Parameters.Select(p => (p.Type, RefKindModifiers: CSharpSyntaxGenerator.GetParameterModifiers(p.RefKind)))
+                    .Concat(SpecializedCollections.SingletonEnumerable((
+                        Type: symbol.Signature.ReturnType,
+                        RefKindModifiers: CSharpSyntaxGenerator.GetParameterModifiers(symbol.Signature.RefKind, forFunctionPointerReturnParameter: true))))
+                    .SelectAsArray(t => SyntaxFactory.Parameter(SyntaxFactory.MissingToken(SyntaxKind.IdentifierToken)).WithModifiers(t.RefKindModifiers).WithType(t.Type.GenerateTypeSyntax()));
```
I can't decide if this is a thing of beauty or a a horrific abuse of our language. :smile:

333fred replied:
Beauty. Definitely beauty.

CyrusNajmabadi commented:
fwiw, i also had my brain break on that localization message.  what about `taking the address of a method` instead of using `&`?

333fred commented:
> fwiw, i also had my brain break on that localization message. what about taking the address of a method instead of using &?

I updated it to `&method group`. I'm not going to change it further at this point. If we get feedback after release we change it then.

AlekseyTs commented on src/Compilers/CSharp/Portable/PublicAPI.Unshipped.txt:
```diff
@@ -122,6 +124,7 @@ static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.UnaryPattern(Microsoft.CodeAn
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ObjectCreationExpressionSyntax.ArgumentList.get -> Microsoft.CodeAnalysis.CSharp.Syntax.ArgumentListSyntax
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ObjectCreationExpressionSyntax.Initializer.get -> Microsoft.CodeAnalysis.CSharp.Syntax.InitializerExpressionSyntax
 *REMOVED*Microsoft.CodeAnalysis.CSharp.Syntax.ObjectCreationExpressionSyntax.NewKeyword.get -> Microsoft.CodeAnalysis.SyntaxToken
+*REMOVED*static Microsoft.CodeAnalysis.CSharp.SyntaxFactory.ParseTypeName(string text, int offset = 0, bool consumeFullText = true) -> Microsoft.CodeAnalysis.CSharp.Syntax.TypeSyntax
```
Shouldn't we keep this API?

AlekseyTs replied:
Or is the only change that we removed default values for parameters? 

---
In reply to: [432672190](https://github.com/dotnet/roslyn/pull/44564#discussion_r432672190) [](ancestors = 432672190)

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:
```diff
@@ -1693,9 +1693,19 @@ public static NameSyntax ParseName(string text, int offset = 0, bool consumeFull
+        /// <summary>
+        /// Parse a TypeNameSyntax node using the grammar rule for type names.
+        /// </summary>
+        public static TypeSyntax ParseTypeName(string text, int offset = 0, ParseOptions? options = null, bool consumeFullText = true)
```
>options [](start = 90, length = 7)

Why is this needed?

AlekseyTs replied:
>ParseTypeName [](start = 33, length = 13)

Please add targeted compiler test for this API

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Syntax/SyntaxNodeFactories.vb:
```diff
@@ -15,6 +15,7 @@ Imports Microsoft.CodeAnalysis.VisualBasic.SyntaxFacts
 Imports InternalSyntax = Microsoft.CodeAnalysis.VisualBasic.Syntax.InternalSyntax
 Imports Microsoft.CodeAnalysis.Syntax
 Imports System.Collections.Immutable
+Imports System.ComponentModel
```
Is this needed?

333fred replied:
Yes, it's for the `EditorBrowsableAttribute`.

---
In reply to: [432677987](https://github.com/dotnet/roslyn/pull/44564#discussion_r432677987) [](ancestors = 432677987)

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Syntax/SyntaxNodeFactories.vb:
```diff
@@ -191,14 +192,25 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
         ''' <param name="offset">The starting offset in the string</param>
-        Public Shared Function ParseTypeName(text As String, Optional offset As Integer = 0, Optional consumeFullText As Boolean = True) As TypeSyntax
-            Using p = New InternalSyntax.Parser(MakeSourceText(text, offset), VisualBasicParseOptions.Default)
+        Public Shared Function ParseTypeName(text As String, Optional offset As Integer = 0, Optional options As ParseOptions = Nothing, Optional consumeFullText As Boolean = True) As TypeSyntax
```
>ParseOptions [](start = 113, length = 12)

Should this be ```VisualBasicParseOptions``` as in ```ParseCompilationUnit``` and other similar APIs?

AlekseyTs replied:
>ParseTypeName [](start = 31, length = 13)

Please add targeted compiler test for this API

AlekseyTs commented on src/Compilers/CSharp/Portable/Syntax/SyntaxFactory.cs:
```diff
@@ -1693,9 +1693,19 @@ public static NameSyntax ParseName(string text, int offset = 0, bool consumeFull
+        /// <summary>
+        /// Parse a TypeNameSyntax node using the grammar rule for type names.
+        /// </summary>
+        public static TypeSyntax ParseTypeName(string text, int offset = 0, ParseOptions? options = null, bool consumeFullText = true)
```
>ParseOptions [](start = 76, length = 12)

Should this be ```CSharpParseOptions``` as in ```ParseCompilationUnit```?

333fred replied:
We seem to be inconsistent as to whether we take a ParseOptions or a CSharpParseOptions, but given that this (and the rest where we take a `ParseOptions`) just casts immediately I'll make a `CSharpParseOptions`.

---
In reply to: [432679991](https://github.com/dotnet/roslyn/pull/44564#discussion_r432679991) [](ancestors = 432679991)

AlekseyTs commented:
Done with review pass for changes under Compilers (iteration 13)

333fred commented:
@AlekseyTs addressed feedback

---
In reply to: [636138129](https://github.com/dotnet/roslyn/pull/44564#issuecomment-636138129) [](ancestors = 636138129)

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs:
```diff
@@ -29,6 +29,13 @@ private CompilationVerifier CompileAndVerifyFunctionPointers(CSharpCompilation c
+        public void UsingTest()
+        {
+            CreateCompilationWithFunctionPointers(@"
+using s = delegate*<void>;").VerifyDiagnostics();
```
I would expect us to try using alias in a meaningful way, compile and run the code. Could be followed up on in a separate PR.

AlekseyTs replied:
It looks like this test is failing in CI, a change is needed anyway.

---
In reply to: [432721087](https://github.com/dotnet/roslyn/pull/44564#discussion_r432721087) [](ancestors = 432721087)
